src/zkbob/ZkBobPool.sol

```diff
@@ -242,6 +307,143 @@ contract ZkBobPool is EIP1967Admin, Ownable, Parameters, ZkBobAccounting {
         }
     }
 
+    function appendDirectDeposits(
+        uint256 _root_after,
+        uint256[] calldata _indices,
+        uint256 _out_commit,
+        uint256[8] memory _batch_deposit_proof,
+        uint256[8] memory _tree_proof
+    )
+        external
+        onlyOperator
+    {
+        require(_indices.length > 0, "ZkBobPool: empty deposit list");
+        require(_indices.length < 17, "ZkBobPool: too many deposits");
+
+        uint256[33] memory batch_deposit_pub;
+        uint256 total = 0;
+        for (uint256 i = 0; i < _indices.length; i++) {
+            DirectDeposit storage dd = directDeposits[_indices[i]];
+            (bytes10 diversifier, uint64 deposit, DirectDepositStatus status) = (dd.diversifier, dd.deposit, dd.status);
+            require(status == DirectDepositStatus.Pending, "ZkBobPool: direct deposit not pending");
+
+            // TODO format
+            batch_deposit_pub[2 * i] = uint256(dd.pk);
+            batch_deposit_pub[2 * i + 1] = uint256(bytes32(diversifier) | bytes32(uint256(deposit)));
+
+            dd.status = DirectDepositStatus.Completed;
+        }
+        batch_deposit_pub[32] = _out_commit;
+
+        uint256 txCount = _processDirectDepositBatch(total);
+        uint256 _pool_index = txCount << 7;
+
+        // verify that _out_commit corresponds to zero output account + 16 chosen notes + 111 empty notes
+        require(
+            batch_deposit_verifier.verifyProof(batch_deposit_pub, _batch_deposit_proof),
+            "ZkBobPool: bad batch deposit proof"
+        );
+
+        uint256[3] memory tree_pub = [roots[_pool_index], _root_after, _out_commit];
+        require(tree_verifier.verifyProof(tree_pub, _tree_proof), "ZkBobPool: bad tree proof");
+
+        _pool_index += 128;
+        roots[_pool_index] = _tree_root_after();
+        bytes memory message; // TODO
+        assembly {
+            message := batch_deposit_pub
+            mstore(message, mul(64, calldataload(_indices.offset)))
+        }
+        bytes32 message_hash = keccak256(message);
+        bytes32 _all_messages_hash = keccak256(abi.encodePacked(all_messages_hash, message_hash));
+        all_messages_hash = _all_messages_hash;
+        emit Message(_pool_index, _all_messages_hash, message);
+        emit CompleteDirectDepositBatch(_pool_index, _indices);
+    }
+
+    function directDeposit(address _fallbackUser, uint256 _amount, bytes memory _rawZkAddress) external {
+        _recordDirectDeposit(_fallbackUser, _amount, _rawZkAddress);
```

Comment:
Looks like a call of `transferFrom` is missed here, doesn't it?

Reply:
indeed, fixed